pkg/operation/botanist/controlplane/resourcemanager/resource_manager.go

```diff
@@ -0,0 +1,553 @@
+
+	appsv1 "k8s.io/api/apps/v1"
+	autoscalingv1 "k8s.io/api/autoscaling/v1"
+	corev1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	autoscalingv1beta2 "k8s.io/autoscaler/vertical-pod-autoscaler/pkg/apis/autoscaling.k8s.io/v1beta2"
+	"k8s.io/utils/pointer"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+const (
+	// SecretName is a constant for the secret name for the gardener resource manager's kubeconfig secret.
+	SecretName = "gardener-resource-manager"
+	// UserName is the name that should be used for the secret that the gardener resource manager uses to
+	// authenticate itself with the kube-apiserver (e.g., the common name in its client certificate).
+	UserName = "gardener-resource-manager"
+
+	containerName      = v1beta1constants.DeploymentNameGardenerResourceManager
+	healthPort         = 8081
+	metricsPort        = 8080
+	roleName           = "gardener-resource-manager"
+	serviceAccountName = "gardener-resource-manager"
+)
+
+var (
+	allowAll = []rbacv1.PolicyRule{{
+		APIGroups: []string{"*"},
+		Resources: []string{"*"},
+		Verbs:     []string{"*"},
+	}}
+
+	allowManagedResources = []rbacv1.PolicyRule{{
+		APIGroups: []string{"resources.gardener.cloud"},
+		Resources: []string{"managedresources", "managedresources/status"},
+		Verbs:     []string{"get", "list", "watch", "update", "patch"},
+	}, {
+		APIGroups: []string{""},
+		Resources: []string{"secrets"},
+		Verbs:     []string{"get", "list", "watch", "update", "patch"},
+	}, {
+		APIGroups: []string{""},
+		Resources: []string{"configmaps", "events"},
+		Verbs:     []string{"create"},
+	}, {
+		APIGroups:     []string{""},
+		Resources:     []string{"configmaps"},
+		ResourceNames: []string{"gardener-resource-manager"},
+		Verbs:         []string{"get", "watch", "update", "patch"},
+	}}
+)
+
+// New creates a new instance of the gardener-resource-manager.
+func New(
+	client client.Client,
+	namespace string,
+	image string,
+	replicas int32,
+	config Config,
+) *ResourceManager {
+	if config.DefaultLabels == nil {
+		config.DefaultLabels = appLabel()
+	} else {
+		defaultLabelsWithAppLabel := utils.MergeStringMaps(*config.DefaultLabels, *appLabel())
+		config.DefaultLabels = &defaultLabelsWithAppLabel
+	}
+	return &ResourceManager{
+		client:    client,
+		image:     image,
+		namespace: namespace,
+		replicas:  replicas,
+		config:    config,
+	}
+}
+
+type ResourceManager struct {
+	client    client.Client
+	namespace string
+	image     string
+	replicas  int32
+	config    Config
+}
+
+// Config holds the optional configuration options for the gardener resource manager
+type Config struct {
+	AlwaysUpdate               *bool
+	ConcurrentSyncs            *int32
+	ClusterRoleName            *string
+	DefaultLabels              *map[string]string
```

Comment:
```suggestion
	Labels              map[string]string
```

`map` is already a pointer type